### DISCUSSION TITLE
All backups are running on a single node

### DISCUSSION THREAD

**Discussion by erichevers** [longhorn/longhorn#8271] [General]
In our cluster all the backups are running on the same single node. We have over 60 pvc's so the node is complaining that it has too many pods running. Is it normal that all the backups only use one node, or is it possible to spread the load over multiple nodes?

**Comment by ChanYiLin** [comment]
Hi @erichevers 
When you said "backup", did you mean "volumes restored from the backups"
Because "backup" in Longhorn means the "backup of the volume that stored in the backupstore(nfs, s3)". It doesn't have any pod running somewhere.

Would you mind describing more about the "backup" you referred to? Thanks!

**Comment by erichevers** [comment]
Hi @ChanYiLin, 
I've scheduled backups of all volumes to S3. When i do "kubectl get pods -n longhorn-system -o wide | grep c-", I see that around 60 pods are Completed in the last 24 hours and they have all been running on the same node. when i do "kubectl logs c-t0dgtz-28533840-jkmvm -n longhorn-system" the logfile indicates it is a longhorn backup pod that has been running and Completed. Is it possible to spread these backup jobs over multiple nodes, instead of a single node?

**Comment by ChanYiLin** [comment]
Hi @erichevers 
I see, you are referring to the RecurringJob
The job periodically create a worker pod to the backup of the volume to the S3.
It will release the resource after the job is done and the pod is completed.

We do have settings to control where it can run
but the settings are shared by all the system managed components
https://github.com/longhorn/longhorn-manager/blob/89b7ae6a8c6d611a2bfcdfb9b19affe8d92e54e2/controller/recurring_job_controller.go#L458-L474

You can find the setting in the longhorn dashboard. Here is the doc: https://longhorn.io/docs/1.6.1/references/settings/#danger-zone
- `System Managed Components Node Selector`
- `Kubernetes Taint Toleration`

If all the jobs run on the same node, maybe that is because these two settings limit the choice of where they can run on.
cc @c3y1huang

**Comment by erichevers** [comment]
Hi @ChanYiLin ,
I checked the Settings, but both are empty
Attached is a screenshot of the settings.
![2024-04-02_19-17-54](https://github.com/longhorn/longhorn/assets/23660309/a6fd61fb-cfc7-40a7-94a9-6193919ad838)

Regards,
Eric

**Comment by ChanYiLin** [comment]
Can you use kubectl to see if those pods have taints or affinities?
You can also check the k8s cronjobs to see if they have any settings
Our recurringJob leverages the cronjob from k8s.
So the pods created by cronjob should be randomly placed on different node.

Could you also please provide more details of your cluster and environment?

Thanks

**Comment by erichevers** [comment]
Hi @ChanYiLin ,
When i do "kubectl get cronjobs --all-namespaces" i only have cronjobs in the longhorn-system namespace. 

"kubectl describe cronjob c-t0dgtz -n longhorn-system" gives:
Name:                          c-t0dgtz
Namespace:                     longhorn-system
Labels:                        longhorn.io/managed-by=longhorn-manager
                               recurring-job.longhorn.io=c-t0dgtz
Annotations:                   longhorn.io/last-applied-cronjob-spec:
                                 {"schedule":"0 6 * * *","concurrencyPolicy":"Forbid","jobTemplate":{"metadata":{"creationTimestamp":null},"spec":{"backoffLimit":3,"templa...
Schedule:                      0 6 * * *
Concurrency Policy:            Forbid
Suspend:                       False
Successful Job History Limit:  1
Failed Job History Limit:      1
Starting Deadline Seconds:     <unset>
Selector:                      <unset>
Parallelism:                   <unset>
Completions:                   <unset>
Pod Template:
  Labels:           longhorn.io/managed-by=longhorn-manager
                    recurring-job.longhorn.io=c-t0dgtz
  Service Account:  longhorn-service-account
  Containers:
   c-t0dgtz:
    Image:      longhornio/longhorn-manager:v1.5.3
    Port:       <none>
    Host Port:  <none>
    Command:
      longhorn-manager
      -d
      recurring-job
      c-t0dgtz
      --manager-url
      http://longhorn-backend:9500/v1
    Environment:
      POD_NAMESPACE:   (v1:metadata.namespace)
    Mounts:
      /var/lib/longhorn/engine-binaries/ from engine-binaries (rw)
  Volumes:
   engine-binaries:
    Type:            HostPath (bare host directory volume)
    Path:            /var/lib/longhorn/engine-binaries/
    HostPathType:    
Last Schedule Time:  Wed, 03 Apr 2024 06:00:00 +0200
Active Jobs:         <none>
Events:              <none>

The pod that was running for this cronjob.
"kubectl describe pod c-t0dgtz-28535280-pdjqx -n longhorn-system" gives:
Name:             c-t0dgtz-28535280-pdjqx
Namespace:        longhorn-system
Priority:         0
Service Account:  longhorn-service-account
Node:             nl1k8s002/192.168.x.x // changed the ip-adres used
Start Time:       Wed, 03 Apr 2024 06:00:00 +0200
Labels:           batch.kubernetes.io/controller-uid=76cbbc13-456b-47e4-866e-4c2ffe907cb5
                  batch.kubernetes.io/job-name=c-t0dgtz-28535280
                  controller-uid=76cbbc13-456b-47e4-866e-4c2ffe907cb5
                  job-name=c-t0dgtz-28535280
                  longhorn.io/managed-by=longhorn-manager
                  recurring-job.longhorn.io=c-t0dgtz
Annotations:      cni.projectcalico.org/containerID: a65042aa8910c21a95c95e1adb6f3109420d9ed664dc1d09f1f429bc3860514e
                  cni.projectcalico.org/podIP: 
                  cni.projectcalico.org/podIPs: 
Status:           Succeeded
IP:               10.1.88.18
IPs:
  IP:           10.1.88.18
Controlled By:  Job/c-t0dgtz-28535280
Containers:
  c-t0dgtz:
    Container ID:  containerd://1bc14131706c264c37090fd037960354b72eeb6652aa8dc76133a85b1fa676db
    Image:         longhornio/longhorn-manager:v1.5.3
    Image ID:      docker.io/longhornio/longhorn-manager@sha256:b5769e761b172b441567c11416c3181078e29dea450683eb1bc66192b0caa494
    Port:          <none>
    Host Port:     <none>
    Command:
      longhorn-manager
      -d
      recurring-job
      c-t0dgtz
      --manager-url
      http://longhorn-backend:9500/v1
    State:          Terminated
      Reason:       Completed
      Exit Code:    0
      Started:      Wed, 03 Apr 2024 06:00:01 +0200
      Finished:     Wed, 03 Apr 2024 06:00:31 +0200
    Ready:          False
    Restart Count:  0
    Environment:
      POD_NAMESPACE:  longhorn-system (v1:metadata.namespace)
    Mounts:
      /var/lib/longhorn/engine-binaries/ from engine-binaries (rw)
      /var/run/secrets/kubernetes.io/serviceaccount from kube-api-access-g7dwz (ro)
Conditions:
  Type              Status
  Initialized       True 
  Ready             False 
  ContainersReady   False 
  PodScheduled      True 
Volumes:
  engine-binaries:
    Type:          HostPath (bare host directory volume)
    Path:          /var/lib/longhorn/engine-binaries/
    HostPathType:  
  kube-api-access-g7dwz:
    Type:                    Projected (a volume that contains injected data from multiple sources)
    TokenExpirationSeconds:  3607
    ConfigMapName:           kube-root-ca.crt
    ConfigMapOptional:       <nil>
    DownwardAPI:             true
QoS Class:                   BestEffort
Node-Selectors:              <none>
Tolerations:                 node.kubernetes.io/not-ready:NoExecute op=Exists for 300s
                             node.kubernetes.io/unreachable:NoExecute op=Exists for 300s
Events:                      <none>

The job that was running.
"kubectl describe job c-t0dgtz-28535280 -n longhorn-system" gives:
Name:             c-t0dgtz-28535280
Namespace:        longhorn-system
Selector:         batch.kubernetes.io/controller-uid=76cbbc13-456b-47e4-866e-4c2ffe907cb5
Labels:           batch.kubernetes.io/controller-uid=76cbbc13-456b-47e4-866e-4c2ffe907cb5
                  batch.kubernetes.io/job-name=c-t0dgtz-28535280
                  controller-uid=76cbbc13-456b-47e4-866e-4c2ffe907cb5
                  job-name=c-t0dgtz-28535280
                  longhorn.io/managed-by=longhorn-manager
                  recurring-job.longhorn.io=c-t0dgtz
Annotations:      batch.kubernetes.io/cronjob-scheduled-timestamp: 2024-04-03T04:00:00Z
Controlled By:    CronJob/c-t0dgtz
Parallelism:      1
Completions:      1
Completion Mode:  NonIndexed
Start Time:       Wed, 03 Apr 2024 06:00:00 +0200
Completed At:     Wed, 03 Apr 2024 06:00:34 +0200
Duration:         34s
Pods Statuses:    0 Active (0 Ready) / 1 Succeeded / 0 Failed
Pod Template:
  Labels:           batch.kubernetes.io/controller-uid=76cbbc13-456b-47e4-866e-4c2ffe907cb5
                    batch.kubernetes.io/job-name=c-t0dgtz-28535280
                    controller-uid=76cbbc13-456b-47e4-866e-4c2ffe907cb5
                    job-name=c-t0dgtz-28535280
                    longhorn.io/managed-by=longhorn-manager
                    recurring-job.longhorn.io=c-t0dgtz
  Service Account:  longhorn-service-account
  Containers:
   c-t0dgtz:
    Image:      longhornio/longhorn-manager:v1.5.3
    Port:       <none>
    Host Port:  <none>
    Command:
      longhorn-manager
      -d
      recurring-job
      c-t0dgtz
      --manager-url
      http://longhorn-backend:9500/v1
    Environment:
      POD_NAMESPACE:   (v1:metadata.namespace)
    Mounts:
      /var/lib/longhorn/engine-binaries/ from engine-binaries (rw)
  Volumes:
   engine-binaries:
    Type:          HostPath (bare host directory volume)
    Path:          /var/lib/longhorn/engine-binaries/
    HostPathType:  
Events:            <none>

I don't see any taints or affinities in any of these

**Comment by ChanYiLin** [comment]
How many nodes in your cluster?
And how many nodes longhorn used?

If there is no taints or affinities then the  scheduling is decided by k8s and should be round robin
I dont know why it only runs on the same node.

**Comment by erichevers** [comment]
Hi @ChanYiLin,
there are 8 nodes in the cluster, all use longhorn, but i've 2 nodes (node 7 + 8) without disks assigned.
Ok, so it is k8s that assigns always the same node. Strange, but let's leave it at that.
Thanks for looking into this.

**Comment by derekbit** [comment]
In this case, the backup jobs are assigned to the same node. They are randomly determined by the engine proxies in the IM pods.
Currently, we cannot do affinity or anti-affinity of the backup jobs.
Convert the question to a feature request.

**Comment by ChanYiLin** [comment]
Hi @erichevers 
Are you still encountering this situation?

**Comment by erichevers** [comment]
Hi Jack,

I’m not running Longhorn anymore. Although I like Longhorn a lot, we needed to move to Ceph for some non-technical reasons.
You can close the case.

Regards,
EriC


Van: Jack Lin ***@***.***>
Datum: donderdag, 15 augustus 2024 om 05:30
Aan: longhorn/longhorn ***@***.***>
CC: EEvers ***@***.***>, Mention ***@***.***>
Onderwerp: Re: [longhorn/longhorn] All backups are running on a single node (Discussion #8271)

Hi @erichevers<https://github.com/erichevers>
Are you still encountering this situation?
